environmental/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+	// `f`.
+	// while we record this pointer (while it's non-zero) we guarantee:
+	// - it will only be used once at any time (no reentrancy);
+	// - that no other thread will use it; and
+	// - that we do not use the original mutating reference while the pointer.
+	// exists.
+	let original = global.with(|r| {
+		let mut b = r.borrow_mut();
+		let o = *b;
+		*b = protected as *mut T as *mut S;
+		o
+	});
+	let r = f();
+	global.with(|r| *r.borrow_mut() = original);
+	r
+}
+
+#[doc(hidden)]
+pub fn with<'r, R, S, T: 'r, F: FnOnce(&'r mut T) -> R>(
+	global: &'static LocalKey<RefCell<*mut S>>,
+	mutator: F,
+) -> Option<R> {
+	global.with(|r| {
+		let br = r.borrow_mut();
+		if *br != 0 as *mut S {
+			// safe because it's only non-zero when it's being called from using, which
+			// is holding on to the underlying reference (and not using it itself) safely.
+			unsafe {
+				Some(mutator(&mut *(*br as *mut S as *mut T)))
+			}
+		} else {
+			None
+		}
+	})
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! decl {
+	($name:ident : $t:ty) => {
+		thread_local! {
+			static $name: $crate::RefCell<*mut $t> = $crate::RefCell::new(0 as *mut $t);
+		}
+	}
+}
+
+/// Declare a new global reference module whose underlying value does not contain references.
+///
+/// Will create a module of a given name that contains two functions:
+/// * `pub fn using<'a: 'b, 'b, R, F: FnOnce() -> R, T: 'a>(protected: &'b mut T, f: F) -> R`
+///   This executes `f`, returning its value. During the call, the module's reference is set to
+///   be equal to `protected`.
+/// * `pub fn with<R, F: for<'r, 't: 'r> FnOnce(&'r mut $t<'t>) -> R>(f: F) -> Option<R>`
+///   This executes `f`, returning `Some` of its value if called from code that is being executed
+///   as part of a `using` call. If not, it returns `None`. `f` is provided with one argument: the
+///   same reference as provided to the most recent `using` call.
+///
+/// # Examples
+///
+/// ```
+/// #[macro_use] extern crate environmental;
+/// declare_simple!(counter: u32);
+/// fn main() {
+///   let mut counter_value = 41u32;
+///   counter::using(&mut counter_value, || {
+///     let odd = counter::with(|value|
+///       if *value % 2 == 1 {
+///         *value += 1; true
+///       } else {
+///         *value -= 3; false
+///       }).unwrap();	// safe because we're inside a counter::using
+///     println!("counter was {}", match odd { true => "odd", _ => "even" });
+///   });
+///
+///   println!("The answer is {:?}", counter_value); // 42
+/// }
+/// ```
+#[macro_export]
+macro_rules! declare_simple {
+	($name:ident : $t:tt) => {
+		mod $name {
+			#[allow(unused_imports)]
+			use super::*;
+
+			decl!(GLOBAL: $t);
+
+			pub fn using<'a: 'b, 'b, R, F: FnOnce() -> R, T: 'a>(
+				protected: &'b mut T,
+				f: F
+			) -> R {
+				$crate::using(&GLOBAL, protected, f)
+			}
+
+			pub fn with<R, F: for<'r> FnOnce(&'r mut $t) -> R>(
```

Comment:
this is completely unsafe:

```rust
	#[test]
	fn break_it() {
		declare_simple!(items: Vec<u8>);

		let mut local_items = vec![1, 2, 3];
		items::using(&mut local_items, || {
			let mut dies_at_end = vec![4, 5, 6];
			items::with(|items| *items = dies_at_end);
		});

        // this compiles and thus local_items points to freed memory
	}
```

Reply:
consider using invariant lifetimes -- worth noting that the simple bound of `'b, 'a: 'b` doesn't actually denote a strictly outlives relationship and allows 'b to be shrunk to the minimum necessary lifetime.

Reply:
for this kind of code we should really have compile-fail tests in the CI: https://crates.io/crates/compiletest_rs